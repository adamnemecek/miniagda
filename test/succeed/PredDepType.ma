sized data Nat : Size -> Set
{ zero : [i : Size] -> Nat ($ i)
; succ : [i : Size] -> Nat i -> Nat ($ i)
}

fun Pred : (i : Size) -> (x : Nat ($ i)) -> Set
{ Pred i (succ .i n) = Nat i
; Pred i (zero .i)   = Nat ($ i)
}

fun pred : [i : Size] -> (x : Nat ($ i)) -> Pred i x
{ pred i (succ .i n) = n
; pred i (zero .i)   = zero i
}

{- DOES NOT WORK
fun minus : [i : Size] -> Nat i -> Nat # -> Nat i
{ minus i n (zero .#) = n
; minus i n (succ .# m) = minus i (pred i n) m 
}
-}