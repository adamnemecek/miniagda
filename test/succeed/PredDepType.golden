MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "PredDepType.ma" ---
--- scope checking ---
--- type checking ---
type  Nat : + Size -> Set
term  Nat.zero : .[s!ze : Size] -> .[i < s!ze] -> Nat s!ze
term  Nat.zero : .[i : Size] -> < Nat.zero i : Nat $i >
term  Nat.succ : .[s!ze : Size] -> .[i < s!ze] -> ^ Nat i -> Nat s!ze
term  Nat.succ : .[i : Size] -> ^(y1 : Nat i) -> < Nat.succ i y1 : Nat $i >
type  Pred : (i : Size) -> (x : Nat $i) -> Set
{ Pred i (Nat.succ [.i] n) = Nat i
; Pred i (Nat.zero [.i]) = Nat $i
}
term  pred : .[i : Size] -> (x : Nat $i) -> Pred i x
{ pred [i] (Nat.succ [.i] n) = n
; pred [i] (Nat.zero [.i]) = Nat.zero [i]
}
--- evaluating ---
--- closing "PredDepType.ma" ---
